File: Intatis-Windows/src/Intatis.Core/Session/ConversationProjection.cs
using System.Text.Json.Nodes;
using Intatis.Core.Protocol;

namespace Intatis.Core.Session;

public sealed record ChatMessageView
{
    public string MessageId { get; init; } = "";
    public MessageRoleWire Role { get; init; }
    public string? Agent { get; init; }
    public string Text { get; init; } = "";
    public bool IsComplete { get; init; }
    public DateTime Timestamp { get; init; }
    public int AttachmentCount { get; init; }
    public List<MessageCitation> Citations { get; init; } = [];
}

/// <summary>
/// Folds envelopes into the chat text view. The UI only ever consumes this folded
/// projection, never raw model output as truth. Unknown event types are ignored.
/// </summary>
public static class ConversationProjection
{
    public static List<ChatMessageView> Build(IEnumerable<Envelope> envelopes)
    {
        var messages = new List<ChatMessageView>();
        ChatMessageView? open = null;
        foreach (var envelope in envelopes)
        {
            switch (envelope.Type)
            {
                case EventType.UserMessage:
                    Flush(ref open, messages);
                    var user = UserMessagePayload.FromJson(envelope.Payload);
                    messages.Add(new ChatMessageView
                    {
                        MessageId = "user_" + envelope.Seq,
                        Role = MessageRoleWire.User,
                        Text = user.Text,
                        IsComplete = true,
                        Timestamp = envelope.Ts,
                        AttachmentCount = user.Attachments?.Count ?? 0,
                    });
                    break;

                case EventType.MessageDelta:
                    var delta = ParseDelta(envelope.Payload);
                    if (open is null || open.MessageId != delta.MessageId)
                    {
                        Flush(ref open, messages);
                        open = new ChatMessageView
                        {
                            MessageId = delta.MessageId,
                            Role = MessageRoleWireExtensions.FromWire(delta.Role),
                            Agent = delta.Agent,
                            Timestamp = envelope.Ts,
                        };
                    }
                    open = open with { Text = open.Text + delta.TextDelta };
                    break;

                case EventType.MessageCompleted:
                    var completed = ParseCompleted(envelope.Payload);
                    if (open is not null && open.MessageId == completed.MessageId)
                        messages.Add(open with { Text = completed.Text, IsComplete = true, Citations = completed.Citations });
                    else
                        messages.Add(new ChatMessageView
                        {
                            MessageId = completed.MessageId,
                            Role = MessageRoleWireExtensions.FromWire(completed.Role),
                            Agent = completed.Agent,
                            Text = completed.Text,
                            IsComplete = true,
                            Timestamp = envelope.Ts,
                            Citations = completed.Citations,
                        });
                    open = null;
                    break;

                case EventType.Error:
                    Flush(ref open, messages);
                    var code = (string?)envelope.Payload?["code"] ?? "error";
                    var message = (string?)envelope.Payload?["message"] ?? "unknown error";
                    messages.Add(new ChatMessageView
                    {
                        MessageId = "error_" + envelope.Seq,
                        Role = MessageRoleWire.System,
                        Text = $"[{code}] {message}",
                        IsComplete = true,
                        Timestamp = envelope.Ts,
                    });
                    break;
            }
        }
        Flush(ref open, messages);
        return messages;
    }

    private static void Flush(ref ChatMessageView? open, List<ChatMessageView> messages)
    {
        if (open is not null) messages.Add(open);
        open = null;
    }

    private static MessageDeltaPayload ParseDelta(JsonNode? node)
    {
        var o = node as JsonObject;
        return new MessageDeltaPayload
        {
            MessageId = (string?)o?["message_id"] ?? "",
            Role = (string?)o?["role"] ?? "assistant",
            Agent = (string?)o?["agent"],
            TextDelta = (string?)o?["text_delta"] ?? "",
        };
    }

    private static MessageCompletedPayload ParseCompleted(JsonNode? node)
    {
        var o = node as JsonObject;
        var citations = new List<MessageCitation>();
        if (o?["citations"] is JsonArray arr)
        {
            foreach (var item in arr)
            {
                if (item is null) continue;
                citations.Add(new MessageCitation
                {
                    Url = (string?)item["url"] ?? "",
                    Title = (string?)item["title"] ?? "",
                });
            }
        }
        return new MessageCompletedPayload
        {
            MessageId = (string?)o?["message_id"] ?? "",
            Role = (string?)o?["role"] ?? "assistant",
            Agent = (string?)o?["agent"],
            Text = (string?)o?["text"] ?? "",
            Citations = citations,
        };
    }
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/session/ConversationProjection.kt
package com.intatis.shared.session

import com.intatis.shared.protocol.Envelope
import java.time.Instant
import com.intatis.shared.protocol.EventType
import com.intatis.shared.protocol.Jsonx.str
import com.intatis.shared.protocol.MessageCitation
import com.intatis.shared.protocol.MessageRole
import com.intatis.shared.protocol.UserMessagePayload

data class ChatMessageView(
    val messageId: String,
    val role: MessageRole,
    val agent: String? = null,
    val text: String = "",
    val isComplete: Boolean = false,
    val timestamp: Instant = java.time.Instant.now(),
    val attachmentCount: Int = 0,
    val citations: List<MessageCitation> = emptyList(),
)

/**
 * Folds envelopes into the chat text view. The UI only ever consumes this folded
 * projection, never raw model output as truth. Unknown event types are ignored.
 */
object ConversationProjection {

    fun build(envelopes: List<Envelope>): List<ChatMessageView> {
        val messages = mutableListOf<ChatMessageView>()
        var open: ChatMessageView? = null

        fun flush() {
            open?.let { messages.add(it) }
            open = null
        }

        for (envelope in envelopes) {
            when (envelope.type) {
                EventType.USER_MESSAGE -> {
                    flush()
                    val user = UserMessagePayload.fromJson(envelope.payload)
                    messages.add(ChatMessageView(
                        messageId = "user_${envelope.seq}",
                        role = MessageRole.USER,
                        text = user.text,
                        isComplete = true,
                        timestamp = envelope.ts,
                        attachmentCount = user.attachments?.size ?: 0,
                    ))
                }

                EventType.MESSAGE_DELTA -> {
                    val payload = envelope.payload
                    val messageId = payload?.str("message_id") ?: ""
                    if (open == null || open!!.messageId != messageId) {
                        flush()
                        open = ChatMessageView(
                            messageId = messageId,
                            role = MessageRole.fromWire(payload?.str("role") ?: "assistant"),
                            agent = payload?.str("agent"),
                            timestamp = envelope.ts,
                        )
                    }
                    open = open!!.copy(text = open!!.text + (payload?.str("text_delta") ?: ""))
                }

                EventType.MESSAGE_COMPLETED -> {
                    val payload = envelope.payload
                    val messageId = payload?.str("message_id") ?: ""
                    val completed = ChatMessageView(
                        messageId = messageId,
                        role = MessageRole.fromWire(payload?.str("role") ?: "assistant"),
                        agent = payload?.str("agent"),
                        text = payload?.str("text") ?: "",
                        isComplete = true,
                        timestamp = envelope.ts,
                        citations = parseCitations(payload?.get("citations") as? kotlinx.serialization.json.JsonArray),
                    )
                    if (open != null && open!!.messageId == messageId) {
                        messages.add(open!!.copy(
                            text = completed.text,
                            isComplete = true,
                            citations = completed.citations,
                        ))
                    } else {
                        messages.add(completed)
                    }
                    open = null
                }

                EventType.ERROR -> {
                    flush()
                    val code = envelope.payload?.str("code") ?: "error"
                    val message = envelope.payload?.str("message") ?: "unknown error"
                    messages.add(ChatMessageView(
                        messageId = "error_${envelope.seq}",
                        role = MessageRole.SYSTEM,
                        text = "[$code] $message",
                        isComplete = true,
                        timestamp = envelope.ts,
                    ))
                }
            }
        }
        flush()
        return messages
    }

    private fun parseCitations(array: kotlinx.serialization.json.JsonArray?): List<MessageCitation> {
        if (array == null) return emptyList()
        return array.mapNotNull { item ->
            val obj = item as? kotlinx.serialization.json.JsonObject ?: return@mapNotNull null
            MessageCitation(url = obj.str("url") ?: "", title = obj.str("title") ?: "")
        }
    }
}
